### PR TITLE
add farming feature for loans.

### PR DESCRIPTION
**Pull request by zhoujia6139**

### CODE REVIEW

File: runtime/kerria/src/lib.rs [hunk 1]
```diff
@@ -517,6 +517,10 @@ impl pallet_assets::Config for Runtime {
     type Extra = ();
 }
 
+parameter_types! {
+    pub const RewardAssetId: CurrencyId = HKO;
```
Comment (yz89) [inline runtime/kerria/src/lib.rs:521]: Use PARA on kerria

File: pallets/loans/src/lib.rs [hunk 2]
```diff
@@ -228,6 +234,18 @@ pub mod pallet {
...
         /// [admin, asset_id]
         UpdatedMarket(Market<BalanceOf<T>>),
+        /// Reward added
+        RewardAdded(T::AccountId, BalanceOf<T>),
+        /// Reward withdrawed
+        RewardWithdrawn(T::AccountId, BalanceOf<T>),
+        /// Event emitted when market reward speed updated.
+        MarketRewardSpeedUpdated(AssetIdOf<T>, BalanceOf<T>),
+        /// Deposited when Reward is distributed to a supplier
+        DistributedSupplierReward(AssetIdOf<T>, T::AccountId, BalanceOf<T>, BalanceOf<T>),
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:244]: should put `Added`, `Updated` ..etc at the end. same to `UpdatedMarket`

File: pallets/loans/src/lib.rs [hunk 3]
```diff
@@ -1011,6 +1209,188 @@ impl<T: Config> Pallet<T> {
...
+            .and_then(|r| r.to_u128())
+            .ok_or(ArithmeticError::Overflow)?;
+        Ok(reward_delta)
+    }
+
+    fn update_reward_supply_index(asset_id: AssetIdOf<T>) -> DispatchResult {
+        let current_block_number = <frame_system::Pallet<T>>::block_number();
+        RewardSupplyState::<T>::try_mutate(asset_id, |supply_state| -> DispatchResult {
+            let delta_block = current_block_number.saturating_sub(supply_state.block);
+            if !delta_block.is_zero() {
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:1249]: quick return to avoid nested scope

File: pallets/loans/src/lib.rs [hunk 4]
```diff
@@ -1011,6 +1209,188 @@ impl<T: Config> Pallet<T> {
...
+                let supply_speed = MarketRewardSpeed::<T>::get(asset_id);
+                if supply_speed > 0 {
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:1251]: should be consistent, maybe `!..is_zero` is better?

File: pallets/loans/src/lib.rs [hunk 5]
```diff
@@ -1011,6 +1209,188 @@ impl<T: Config> Pallet<T> {
...
+
+    fn update_reward_borrow_index(asset_id: AssetIdOf<T>) -> DispatchResult {
+        let current_block_number = <frame_system::Pallet<T>>::block_number();
+        RewardBorrowState::<T>::try_mutate(asset_id, |borrow_state| -> DispatchResult {
+            let delta_block = current_block_number.saturating_sub(borrow_state.block);
+            if !delta_block.is_zero() {
+                let borrow_speed = MarketRewardSpeed::<T>::get(asset_id);
+                if borrow_speed > 0 {
+                    let current_borrow_amount = TotalBorrows::<T>::get(asset_id);
+                    let delta_index = if current_borrow_amount > 0 {
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:1278]: same as above, can this if be inside

File: pallets/loans/src/lib.rs [hunk 6]
```diff
@@ -1011,6 +1209,188 @@ impl<T: Config> Pallet<T> {
...
+
+        Ok(())
+    }
+
+    fn pay_reward(user: &T::AccountId) -> DispatchResult {
+        let pool_account = Self::account_id();
+        let reward_asset = T::RewardAssetId::get();
+        RewardAccured::<T>::try_mutate(user, |total_reward| -> DispatchResult {
+            let total = *total_reward;
+            if total > 0 {
```
Comment (GopherJ) [inline pallets/loans/src/lib.rs:1384]: not necessary to have this `total` alias

File: pallets/loans/src/benchmarking.rs [hunk 7]
```diff
@@ -57,6 +57,14 @@ fn transfer_initial_balance<
     caller: T::AccountId,
 ) {
     let account_id = T::Lookup::unlookup(caller.clone());
+    pallet_assets::Pallet::<T>::force_create(
```
Comment (yz89) [inline pallets/loans/src/benchmarking.rs:60]: HKO is the native token that is managed by the Balance module.

File: pallets/loans/src/benchmarking.rs [hunk 8]
```diff
@@ -102,6 +110,7 @@ fn transfer_initial_balance<
         true,
     )
     .ok();
+    T::Assets::mint_into(HKO, &caller, INITIAL_AMOUNT.into()).unwrap();
```
Comment (yz89) [inline pallets/loans/src/benchmarking.rs:113]: Refer to this: https://github.com/parallel-finance/parallel/blob/master/pallets/loans/src/mock.rs#L277

File: pallets/loans/src/lib.rs (hunk 2 above)
Comment (zhoujia6139) [reply pallets/loans/src/lib.rs:244]: What do you mean by at the end. RewardAdded -> AddedReward?  @GopherJ

File: pallets/loans/src/lib.rs (hunk 3 above)
Comment (zhoujia6139) [reply pallets/loans/src/lib.rs:1249]: ok

File: pallets/loans/src/lib.rs (hunk 4 above)
Comment (zhoujia6139) [reply pallets/loans/src/lib.rs:1251]: ok

File: pallets/loans/src/lib.rs (hunk 6 above)
Comment (zhoujia6139) [reply pallets/loans/src/lib.rs:1384]: total is used by deposit_event.

File: pallets/loans/src/lib.rs [hunk 9]
```diff
@@ -1011,6 +1209,188 @@ impl<T: Config> Pallet<T> {
...
+        RewardSupplyState::<T>::try_mutate(asset_id, |supply_state| -> DispatchResult {
+            let delta_block = current_block_number.saturating_sub(supply_state.block);
+            if !delta_block.is_zero() {
+                let supply_speed = MarketRewardSpeed::<T>::get(asset_id);
+                if supply_speed > 0 {
+                    let total_supply = TotalSupply::<T>::get(asset_id);
+                    let delta_index = if total_supply > 0 {
+                        Self::calculate_reward_delta_index(delta_block, supply_speed, total_supply)?
+                    } else {
+                        0_u128
```
[the remark this answers is not in the document]
Comment (zhoujia6139) [reply pallets/loans/src/lib.rs:1256]: ok

File: pallets/loans/src/lib.rs (hunk 5 above)
Comment (zhoujia6139) [reply pallets/loans/src/lib.rs:1278]: ok

File: pallets/loans/src/benchmarking.rs (hunk 7 above)
Comment (zhoujia6139) [reply pallets/loans/src/benchmarking.rs]: ok.

File: pallets/loans/src/lib.rs [hunk 10]
```diff
@@ -547,6 +625,118 @@ pub mod pallet {
...
+        #[transactional]
+        pub fn add_reward(
+            origin: OriginFor<T>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            ensure!(!amount.is_zero(), Error::<T>::InvalidAmount);
+
+            let reward_asset = T::RewardAssetId::get();
+            let pool_account = Self::account_id();
```
Comment (yz89) [inline pallets/loans/src/lib.rs:641]: We should use another pool account for this one is already used by the HKO market.

File: pallets/loans/src/lib.rs [hunk 11]
```diff
@@ -547,6 +625,118 @@ pub mod pallet {
...
+        pub fn withdraw_missing_reward(
+            origin: OriginFor<T>,
+            target_account: <T::Lookup as StaticLookup>::Source,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            T::UpdateOrigin::ensure_origin(origin)?;
+            ensure!(!amount.is_zero(), Error::<T>::InvalidAmount);
+
+            let reward_asset = T::RewardAssetId::get();
+            let pool_account = Self::account_id();
```
Comment (yz89) [inline pallets/loans/src/lib.rs:667]: Same here

**Review by yz89 — APPROVED** [review]

**Pull request merged**